SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityUpstreamChecker.java

```diff
@@ -0,0 +1,157 @@
+  }
+
+  private List<UpstreamInfo> getExtraUpstreams(Collection<UpstreamInfo> upstreamsInBaragonForRequest, Collection<UpstreamInfo> upstreamsInSingularityForRequest) {
+    for (UpstreamInfo upstreamInSingularity : upstreamsInSingularityForRequest) {
+      final Collection<UpstreamInfo> matches = getEqualUpstreams(upstreamInSingularity, upstreamsInBaragonForRequest);
+      upstreamsInBaragonForRequest.removeAll(matches);
+    }
+    return new ArrayList<>(upstreamsInBaragonForRequest);
+  }
+
+  private SingularityLoadBalancerUpdate syncUpstreamsForService(SingularityRequest singularityRequest, SingularityDeploy deploy, Optional<String> loadBalancerUpstreamGroup) throws InterruptedException, ExecutionException, TimeoutException, IOException {
+    final String singularityRequestId = singularityRequest.getId();
+    final LoadBalancerRequestId loadBalancerRequestId = new LoadBalancerRequestId(singularityRequestId, LoadBalancerRequestType.REMOVE, Optional.absent());
+    Collection<UpstreamInfo> upstreamsInBaragonForRequest = lbClient.getLoadBalancerUpstreamsForRequest(loadBalancerRequestId.toString());
+    Collection<UpstreamInfo> upstreamsInSingularityForRequest = getUpstreamsFromActiveTasksForRequest(singularityRequestId, loadBalancerUpstreamGroup);
+    final List<UpstreamInfo> extraUpstreams = getExtraUpstreams(upstreamsInBaragonForRequest, upstreamsInSingularityForRequest);
+    return lbClient.makeAndSendLoadBalancerRequest(loadBalancerRequestId, Collections.emptyList(), extraUpstreams, deploy, singularityRequest);
+  }
+
+  private boolean noPendingDeploy() {
+    return deployManager.getPendingDeploys().size() == 0;
+  }
+
+  public void doSyncUpstreamForService(SingularityRequest singularityRequest) {
+    if (singularityRequest.isLoadBalanced() && noPendingDeploy()) {
+      final String singularityRequestId = singularityRequest.getId(); //TODO: lock on the requestId
+      final Optional<String> maybeDeployId = deployManager.getInUseDeployId(singularityRequestId);
+      if (maybeDeployId.isPresent()) {
+        final String deployId = maybeDeployId.get();
+        final Optional<SingularityDeploy> maybeDeploy = deployManager.getDeploy(singularityRequestId, deployId);
+        if (maybeDeploy.isPresent()) {
+          final SingularityDeploy deploy = maybeDeploy.get();
+          final Optional<String> loadBalancerUpstreamGroup = deploy.getLoadBalancerUpstreamGroup();
+          final SingularityLoadBalancerUpdate syncUpstreamsUpdate;
+          try {
+            syncUpstreamsUpdate = syncUpstreamsForService(singularityRequest, deploy, loadBalancerUpstreamGroup);
+            final LoadBalancerRequestId loadBalancerRequestId = syncUpstreamsUpdate.getLoadBalancerRequestId();
+            final BaragonRequestState syncUpstreamsState = lbClient.getState(loadBalancerRequestId).getLoadBalancerState();
+            switch (syncUpstreamsState) { //TODO: discuss what to do when the state is not SUCCESS, apart from logging
```

Comment:
documenting from slack conversation:

the immediate response when you first post it will be `WAITING`
I’d imagine we’d want something like:
- POST it (should be waiting after that)
- Poll until no longer in WAITING (or until some timeout)
- if `SUCCESS` then :thumbsup: , if not then maybe we LOG.error (which will also throw up a sentry that notifies us)